STM32F1: M43 PINS_DEBUGGING

tpruvot opened the pull request:
Sample output : 
```
> Send: M43
> Recv: PIN: 0 PA0 (A 0) TEMP_0_PIN protected
> Recv: PIN: 1 PA1 (A 1) TEMP_BED_PIN protected
> Recv: PIN: 2 PA2 <unused/unknown> Output = 1 TIM2 CH3
> Recv: PIN: 3 PA3 <unused/unknown> Input = 1 TIM2 CH4
> Recv: PIN: 4 PA4 SS_PIN Output = 1
> Recv: PIN: 5 PA5 DOGLCD_SCK Output = 0
> Recv: . SCK_PIN Output = 0
> Recv: PIN: 6 PA6 MISO_PIN Input = 1 TIM3 CH1
> Recv: PIN: 7 PA7 DOGLCD_MOSI TIM3 CH2
> Recv: . MOSI_PIN
> Recv: PIN: 8 PA8 HEATER_BED_PIN protected
> Recv: PIN: 9 PA9 TXD TIM1 CH2
> Recv: PIN: 10 PA10 RXD Input = 1 TIM1 CH3
> Recv: PIN: 11 PA11 <unused/unknown> Input = 1
> Recv: PIN: 12 PA12 <unused/unknown> Input = 1
> Recv: PIN: 13 PA13 <unused/unknown> Input = 1
> Recv: PIN: 14 PA14 <unused/unknown> Input = 0
> Recv: PIN: 15 PA15 FAN_PIN protected
> Recv: PIN: 16 PB0 BTN_ENC Input = 1 TIM3 CH3
> Recv: PIN: 17 PB1 <unused/unknown> Input = 1 TIM3 CH4
> Recv: PIN: 18 PB2 <unused/unknown> Input = 1
> Recv: PIN: 19 PB3 X_DIR_PIN protected
> Recv: PIN: 20 PB4 X_STEP_PIN protected
> Recv: PIN: 21 PB5 X_ENABLE_PIN protected
> Recv: PIN: 22 PB6 Y_DIR_PIN protected
> Recv: PIN: 23 PB7 Y_STEP_PIN protected
> Recv: PIN: 24 PB8 Y_ENABLE_PIN protected
> Recv: PIN: 25 PB9 Z_DIR_PIN protected
> Recv: PIN: 26 PB10 <unused/unknown> Input = 1
> Recv: PIN: 27 PB11 <unused/unknown> Input = 1
> Recv: PIN: 28 PB12 <unused/unknown> Output = 1
> Recv: PIN: 29 PB13 <unused/unknown> Output = 0
> Recv: PIN: 30 PB14 <unused/unknown> Output = 0
> Recv: PIN: 31 PB15 <unused/unknown> Input = 1
> Recv: PIN: 32 PC0 FIL_RUNOUT_PIN Input = 0
> Recv: PIN: 33 PC1 X_MIN_PIN protected
> Recv: PIN: 34 PC2 LED_PIN Output = 0
> Recv: PIN: 35 PC3 <unused/unknown> Input = 1
> Recv: PIN: 36 PC4 <unused/unknown> Output = 1
> Recv: PIN: 37 PC5 <unused/unknown> Input = 1
> Recv: PIN: 38 PC6 <unused/unknown> Input = 1 TIM8 CH1
> Recv: PIN: 39 PC7 <unused/unknown> Input = 1 TIM8 CH2
> Recv: PIN: 40 PC8 <unused/unknown> Output = 1 TIM8 CH3
> Recv: PIN: 41 PC9 <unused/unknown> Output = 1 TIM8 CH4
> Recv: PIN: 42 PC10 <unused/unknown> Output = 1
> Recv: PIN: 43 PC11 <unused/unknown> Output = 1
> Recv: PIN: 44 PC12 <unused/unknown> Output = 0
> Recv: PIN: 45 PC13 <unused/unknown> Input = 1
> Recv: PIN: 46 PC14 <unused/unknown> Input = 1
> Recv: PIN: 47 PC15 Y_MIN_PIN protected
> Recv: PIN: 48 PD0 <unused/unknown> Output = 0
> Recv: PIN: 49 PD1 <unused/unknown> Output = 0
> Recv: PIN: 50 PD2 <unused/unknown> Output = 1
> Recv: PIN: 51 PD3 HEATER_0_PIN protected
> Recv: PIN: 52 PD4 <unused/unknown> Output = 1
> Recv: PIN: 53 PD5 <unused/unknown> Output = 1
> Recv: PIN: 54 PD6 <unused/unknown> Input = 1
> Recv: PIN: 55 PD7 <unused/unknown> Output = 1
> Recv: PIN: 56 PD8 <unused/unknown> Output = 0
> Recv: PIN: 57 PD9 <unused/unknown> Output = 0
> Recv: PIN: 58 PD10 <unused/unknown> Output = 0
> Recv: PIN: 59 PD11 <unused/unknown> Output = 1
> Recv: PIN: 60 PD12 LCD_BACKLIGHT_PIN Output = 1
> Recv: PIN: 61 PD13 SERVO0_PIN Output = 0
> Recv: PIN: 62 PD14 <unused/unknown> Output = 0
> Recv: PIN: 63 PD15 <unused/unknown> Output = 0
> Recv: PIN: 64 PE0 Z_STEP_PIN protected
> Recv: PIN: 65 PE1 Z_ENABLE_PIN protected
> Recv: PIN: 66 PE2 E0_DIR_PIN protected
> Recv: PIN: 67 PE3 E0_STEP_PIN protected
> Recv: PIN: 68 PE4 E0_ENABLE_PIN protected
> Recv: PIN: 69 PE5 <unused/unknown> Input = 1
> Recv: PIN: 70 PE6 Z_MIN_PIN protected
> Recv: PIN: 71 PE7 <unused/unknown> Output = 0
> Recv: PIN: 72 PE8 <unused/unknown> Output = 0
> Recv: PIN: 73 PE9 <unused/unknown> Output = 0
> Recv: PIN: 74 PE10 <unused/unknown> Output = 0
> Recv: PIN: 75 PE11 <unused/unknown> Output = 0
> Recv: PIN: 76 PE12 <unused/unknown> Output = 0
> Recv: PIN: 77 PE13 <unused/unknown> Output = 0
> Recv: PIN: 78 PE14 <unused/unknown> Output = 0
> Recv: PIN: 79 PE15 <unused/unknown> Output = 0
> Recv: ok
```

thisiskeithb commented:
This would be handy!

Bob-the-Kuhn commented:
Good news.

Does it report analog input values?

tpruvot commented:
its not finished... but it was ;) problem, analog pins are set "protected" if NUM_ANALOG_INPUTS is set

else: 
Recv: PIN: 0 PA0 (A 0) <unused/unknown> Analog in = 2075 Input = 0
Recv: PIN: 1 PA1 (A 1) <unused/unknown> Analog in = 3640 Input = 0

tpruvot commented:
To note: the wrapper of analogWrite() is using a different pin_t type... may cause build issues on some STM32F1 using analogWrite() calls (on some optional features) 

It's due to the undefined enum type... see endstops.cpp for the fix to apply... unfortunately there is no real/common HAL functions for those

tpruvot commented:
analogRead() may need the same kind of protection

maple code :
```cpp
/* Unlike Wiring and Arduino, this assumes that the pin's mode is set
 * to INPUT_ANALOG. That's faster, but it does require some extra work
 * on the user's part. Not too much, we think ;). */
uint16 analogRead(uint8 pin) {
    adc_dev *dev = PIN_MAP[pin].adc_device;
    if (dev == NULL) {
        return 0;
    }
    return adc_read(dev, PIN_MAP[pin].adc_channel);
}

```

tpruvot commented:
ok, finished for today, i let you check what is the best to do with analogWrite() to avoid chip crashes like the 
 M42 led toggle was doing...

tpruvot commented:
rebased, with PWM/timers pins addition

Phr3d13 commented:
What is this waiting for? This would be really helpful.

Phr3d13 commented:
I can confirm that this works for the stm32f103vet6 chip on the GTM32_PRO_VB (with the noted exception of analog pins being protected).

tpruvot committed:
STM32F1: handle M43 pin list (PINS_DEBUGGING)

+ avoid M42 analogWrite and M43 analogRead crash

libmaple code forces the pin mode change to PWM even if not possible, which crash

void analogWrite(uint8 pin, int duty_cycle8) {
  pinMode(pin,PWM);
  pwmWrite(pin,duty_cycle8 * 257);// 257 maps 255 to 65535 (i.e 255*257 = 65535)
}

+ show PWM capable pins in M43
+ RGB led analogWrite fix
+ sprintf warning fix in print_port

tpruvot commented:
rebased/squashed in one single commit with RGB led analogWrite workaround (seen on SKRmini)

Bob-the-Kuhn commented:
How generic is this pinsDebug.h?  

I'm wondering if it would work for any STM32F board compiled using the STM32generic library/platform.  

If that is the case then maybe it should reside in the HAL_STM32 folder with a `#ifdef STM32GENERIC` to keep it from being used with ST's version of the library.

Phr3d13 commented:
If it was moved into the HAL_STM32 folder, I wouldn't be able to utilize it. In order to compile for my STM32F1 board, I have to move the HAL_STM32 folder outside the scope of the compiler. If I don't, I get a lot of compiler errors and compile failure.

Bob-the-Kuhn commented:
Are you using PlatformIO or Arduino?

I just got M43 working for STM32F407 on PlatformIO.  I'm going to see if it also works for an STM32F1xx board.

Phr3d13 commented:
Arduino, I can't get as reliable builds with platformio as I can arduino.

Bob-the-Kuhn commented:
Please provide the URL for the STM32GENERIC library you are using with Arduino.  I've tried two with bad results.

Phr3d13 commented:
https://github.com/rogerclarkmelbourne/Arduino_STM32 This one is the one I use. But, like noted earlier. I have to delete/move the HAL_STM32 folder from Marlin to compile without watchdog errors. Then I pick the proper board based on my board variant. 
![image](https://user-images.githubusercontent.com/8862382/59576906-72aa5780-908f-11e9-8c82-800b25a501d1.png)

Phr3d13 commented:
I can get close to working builds with platformio on VSCode, but with that, as soon as I push the encoder button on my LCD, the firmware freezes until I restart the machine. I've tried various STM32 lib versions, but they all produce results similar to each other. I haven't tried the 'workaround' that makes Marlin compile in Arduino for me. Maybe I'll try that tomorrow.

Bob-the-Kuhn commented:
I've just created PR #14309 which, in part, uses your pinsDebug.h file.  I hope I've given you enough credit.

One of the purposes of the new PR is to see if your code can be used with other boards and within PlatformIO.

One sour note so far.  I tried to compile a STM32F407 board using Arduino and your file.  Got an error saying that DUE wasn't supported.  Very strange.  It'll probably be several days before I can look into that further.  Chances are excellent it's due to my lack of knowledge about compiling for specific boards.

tpruvot commented:
yes... but this PR includes a few other fixes i detected while playing with M42/43

Bob-the-Kuhn commented:
My PR doesn't replace this one.  It's an experiment to see how generic our M43 routines are.

My expectation is this PR will get merged soon.  In parallel I'll continue to play with my PR.  

>  move the HAL_STM32 folder outside the scope of the compiler

Is this still required?  If yes then I expect it will need to be fixed before this PR can be merged.  But I'm still trying to understand this strange STM32 world so I could be wrong.

>  'workaround' that makes Marlin compile in Arduino

I haven't stumbled across that yet.  Please point me to it.

Bob-the-Kuhn commented:
My PR doesn't replace this one.  It's an experiment to see how generic our M43 routines are.

My expectation is this PR will get merged soon.  In parallel I'll continue to play with my PR.  

>  move the HAL_STM32 folder outside the scope of the compiler

Is this still required?  If yes then I expect it will need to be fixed before this PR can be merged.  But I'm still trying to understand this strange STM32 world so I could be wrong.

>  'workaround' that makes Marlin compile in Arduino

I haven't stumbled across that yet.  Please point me to it.

Bob-the-Kuhn commented on Marlin/src/HAL/HAL_STM32F1/pinsDebug.h:
```diff
@@ -1 +1,125 @@
+#define pwm_status(pin) PWM_PIN(pin)
+#define digitalRead_mod(p) extDigitalRead(p)
+#define NAME_FORMAT(p) PSTR("%-##p##s")
+#define PRINT_PIN(p) do {sprintf_P(buffer, PSTR("%2hd "), int16_t(p)); SERIAL_ECHO(buffer);} while (0)
```
I just tried it on the MKS Robin.  It worked but the formatting was off because this board has 111 pins defined.

There are three mods needed to make it pretty when there are more than 99 pins

On line 39 change ` PSTR("%2hd ")` to ` PSTR("%3hd ")`

Bob-the-Kuhn commented on Marlin/src/HAL/HAL_STM32F1/pinsDebug.h:
```diff
@@ -1 +1,125 @@
+#define PRINT_PIN(p) do {sprintf_P(buffer, PSTR("%2hd "), int16_t(p)); SERIAL_ECHO(buffer);} while (0)
+#define PRINT_PORT(p) print_port(p)
+#define PRINT_ARRAY_NAME(x)  do {sprintf_P(buffer, PSTR("%-" STRINGIFY(MAX_NAME_LENGTH) "s"), pin_array[x].name); SERIAL_ECHO(buffer);} while (0)
+#define MULTI_NAME_PAD 14 // space needed to be pretty if not first name assigned to a pin
```
On line 42 change the 14 to 20

Bob-the-Kuhn commented on Marlin/src/HAL/HAL_STM32F1/pinsDebug.h:
```diff
@@ -1 +1,125 @@
+  #endif
+  */
+  sprintf_P(buffer, PSTR("P%c%hd "), port, int16_t(PIN_MAP[pin].gpio_bit));
+  SERIAL_ECHO(buffer);
```
Between line 122 and 123 add this line.  It keeps the port number the same number of characters in all cases.

if (int16_t(PIN_MAP[pin].gpio_bit) < 10) SERIAL_CHAR(' ');

thinkyhead committed:
3 digit rj pin numbers

thinkyhead committed:
Other suggested changes
